### PR TITLE
Recursive nondet init of pointers to structs [blocks: #3443]

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/java_object_factory_parameters.h
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_object_factory_parameters.h:14]: Why this if they are the same?

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (peterschrammel) [inline src/ansi-c/c_nondet_symbol_factory.cpp]: Why can't this use util/fresh_symbol?

File: src/cbmc/cbmc_parse_options.cpp
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: references -> pointers

File: jbmc/src/java_bytecode/java_object_factory_parameters.h
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_object_factory_parameters.h]: should be squashed into commit that caused this

File: jbmc/src/java_bytecode/java_object_factory_parameters.cpp
Comment (hannes-steffenhagen-diffblue) [inline jbmc/src/java_bytecode/java_object_factory_parameters.cpp:9]: Why is this in a commit called "move object factory parameters to util"?

File: src/cbmc/cbmc_parse_options.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/cbmc/cbmc_parse_options.cpp]: We should have some more detailed description of how this behaves in our docs

File: src/langapi/language_file.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/langapi/language_file.cpp]: This change seems unrelated to "Pass object factory parameters through to object factories"?

File: regression/cbmc/pointer-function-parameters-3/main.c
Comment (hannes-steffenhagen-diffblue) [inline regression/cbmc/pointer-function-parameters-3/main.c]: Could the tests get better names than "pointer-function-parameters-{3,4,5,6}"?

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
Can't see anything wrong with this, but I'll need some more time to look over this before I can approve

File: jbmc/src/java_bytecode/java_object_factory_parameters.h
Comment (danpoe) [reply jbmc/src/java_bytecode/java_object_factory_parameters.h:14]: This would be a place to add options in the future that are not common to both the C and Java object factory. For C at least we expect some special options (e.g., to indicate which length function argument corresponds to which array function arguments).

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (danpoe) [reply src/ansi-c/c_nondet_symbol_factory.cpp]: Yes, this should probably be replaced by a call to `get_fresh_aux_symbol()` since it's called in one place only.

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (xbauch) [inline src/ansi-c/c_nondet_symbol_factory.cpp:134]: So the `depth` is the minimal length of dereference-chain before null-pointer initialisation? And every chain longer than `depth` will have pointers initialised to null but only when they point to a `struct` that appeared somewhere before on this chain?

Comment (danpoe) [reply src/ansi-c/c_nondet_symbol_factory.cpp:134]: Yes, `max_nondet_tree_depth` gives the depth of a chain such that a pointer to a struct will be initialised to `null` if the `depth` of the chain exceeds it and the struct appeared already before somewhere on the chain. If there is no recursion the chains can have any depth.

File: jbmc/src/java_bytecode/java_object_factory_parameters.cpp
Comment (danpoe) [reply jbmc/src/java_bytecode/java_object_factory_parameters.cpp:9]: Ok, renamed the commit to "Move object factory parameters to util and add java object factory parameters" (plus a body with some more details).

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (xbauch) [reply src/ansi-c/c_nondet_symbol_factory.cpp:134]: I see. So it may turn out to be non-terminating but it would have nothing to do with this new code. I was just thinking if there is a more accurate way to describe the input option.

**Review by xbauch — APPROVED** [review]

File: src/ansi-c/c_nondet_symbol_factory.cpp
Comment (danpoe) [reply src/ansi-c/c_nondet_symbol_factory.cpp:134]: I think if there's no recursion then the chains would always have finite length thus it can never be non-terminating. The current behaviour is similar to that of the java object factory.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 9d8736efdd60cfd67f612961766980ea8bc1e250).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90086978
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/cbmc/cbmc_parse_options.cpp
Comment (danpoe) [reply src/cbmc/cbmc_parse_options.cpp]: I've updated the documentation in `object_factory_parameters.h` for the `max_nondet_tree_depth` parameter.

File: regression/cbmc/pointer-function-parameters-3/main.c
Comment (danpoe) [reply regression/cbmc/pointer-function-parameters-3/main.c]: Ok, renamed.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: fed5e30a10def4abbc7d881eaa239c52f53de643).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90271526
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 465286d8cb33c9be1390f88bc3ca70725cd19e2a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90273939
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by thk123 — DISMISSED** [review]
This will need a real TG bump - will create later today.

**Comment by martin-cs** [comment]
@danpoe No chance of an over-view of what this code does is there?

**Comment by thk123** [comment]
Created TG bump: https://github.com/diffblue/test-gen/pull/2444

File: src/cbmc/cbmc_parse_options.cpp
Comment (hannes-steffenhagen-diffblue) [reply src/cbmc/cbmc_parse_options.cpp]: That helps, but it's not user visible, and even for us it's going to take some digging to find out that this parameter does end up there if we ever need to find out

(This is a reply to one of @danpoe 's other comments, but doesn't show up as such in github. Thanks github).

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
I feel user facing documentation is still a bit lacking considering this adds new behaviour.

**Comment by peterschrammel** [comment]
> Created TG bump: diffblue/test-gen#2444

The bump is passing. @thk123, do you approve?

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 233d925a64260391328265348c57a43502197990).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91063140
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/cbmc/cbmc_parse_options.cpp
Comment (chrisr-diffblue) [reply src/cbmc/cbmc_parse_options.cpp]: Definitely agree that this could do with some user documentation and not just source code documentation.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 60fc8c6756f5a3bf863083646312723879dec0ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91899001
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by danpoe** [comment]
@tautschnig @peterschrammel @thk123 This still requires code owner approval if you get a chance.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 9049aff762de6cb03b59d30dd617ce749b2db85c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92377593
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by peterschrammel — APPROVED** [review]
This is ready to go. @allredj  confirmed that there is a passing bump.

**Comment by allredj** [comment]
(bump is here: https://github.com/diffblue/test-gen/pull/2444)

File: src/util/object_factory_parameters.h
Comment (tautschnig) [inline src/util/object_factory_parameters.h]: `= default;`

File: src/util/nondet_bool.h
Comment (tautschnig) [inline src/util/nondet_bool.h]: Let's not make this optional: there is a single call site (and the next commit adds a second one), which could trivially pass in a proper source location.

File: src/cbmc/cbmc_parse_options.cpp
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: This option is not added in the header file.

File: src/cbmc/cbmc_parse_options.h
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.h]: Use `OPT...` instead of the raw string in here.

File: src/cbmc/cbmc_parse_options.cpp
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: Define a `HELP_...` macro.

File: src/langapi/language.h
Comment (tautschnig) [inline src/langapi/language.h]: Whitespace change only?

File: src/ansi-c/ansi_c_language.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_language.h:37]: Added but never invoked?

File: jbmc/unit/java_bytecode/java_object_factory/gen_nondet_string_init.cpp
Comment (tautschnig) [inline jbmc/unit/java_bytecode/java_object_factory/gen_nondet_string_init.cpp:49]: Should be merged into the earlier commit introducing `java_object_factory_parameterst`.

File: scripts/delete_failing_smt2_solver_tests
Comment (tautschnig) [inline scripts/delete_failing_smt2_solver_tests:163]: Please merge into the commit introducing them.

File: regression/cbmc/pointer-function-parameters-struct-non-recursive/test.desc
Comment (tautschnig) [inline regression/cbmc/pointer-function-parameters-struct-non-recursive/test.desc:5]: Add `^EXIT=10$`

**Review by tautschnig — APPROVED** [review]
Some smaller requests remain, otherwise looks good. In future, I would really appreciate if such PRs could 1) be broken into several smaller PRs to make sure reviewing is also possible other than on weekends (where longer consecutive stretches of time may be available) and 2) each commit message has a body.

File: src/ansi-c/ansi_c_language.h
Comment (danpoe) [reply src/ansi-c/ansi_c_language.h:37]: This would e.g. be called by `initialize_goto_model()` where `languaget &` might reference `ansi_c_languaget`.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0b605ea706986409167e95ab9cfc7005d0279007).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92589920
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**
